syntax = "proto3";

package swim;

message NodeInfo {
  string node_id = 1;
  string address = 2;
}

message PingRequest {
  NodeInfo sender = 1;
  string target_node_id = 2;
}

message PingResponse {
  bool alive = 1;
  string message = 2;
}

message PingReqRequest {
  NodeInfo requester = 1;
  string target_node_id = 2;
}

message PingReqResponse {
  bool alive = 1;
  string message = 2;
}

message DisseminationMessage {
  string failed_node_id = 1;
}

message JoinRequest {
  NodeInfo new_node = 1;
}

message JoinResponse {
  repeated NodeInfo membership_list = 1;
}

message MembershipUpdateRequest {
  repeated NodeInfo membership_list = 1;
}

message Empty {}

service FailureDetector {
  rpc Ping(PingRequest) returns (PingResponse);
  rpc IndirectPing(PingReqRequest) returns (PingReqResponse);
  rpc UpdateMembership(MembershipUpdateRequest) returns (Empty);
}

service Dissemination {
  rpc Disseminate(DisseminationMessage) returns (Empty);
  rpc Join(JoinRequest) returns (JoinResponse);
}
